fix: use node's lookup algorithm to find node modules

lennyburdette opened the pull request:
assuming that the package is inside the project root fails when `npm link`ing or using yarn workspaces.

addresses #23

Commit:
fix: use node's lookup algorithm to find node modules

assuming that the package is inside the project root fails when `npm link`ing or using yarn workspaces.

Turbo87 commented:
@lennyburdette I'm not sure this fix is correct. shouldn't you also (or exclusively) need to update the `customResolveOptions`?

lennyburdette commented:
@Turbo87 good question and I'm definitely not confident about the answer!

I chose this strategy because it closely mirrors what [app.import does](https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/ember-app.js#L1450). But I didn't notice the resolve options.

I *think* that we should actually drop `customResolveOptions` after ensuring that the entry point location is correct. 

I'm pretty sure `moduleDirectory` is intended to be a relative path not an absolutely path. It defaults to ["node_modules"](https://github.com/browserify/resolve/blob/75108bc1d196f0cf770c3c7777f349cc4c3373b2/lib/node-modules-paths.js#L8) and is appended to lookup paths [here](https://github.com/browserify/resolve/blob/75108bc1d196f0cf770c3c7777f349cc4c3373b2/lib/node-modules-paths.js#L40).

The entry point can be absolute—it's actually converted to an absolute path [here](https://github.com/browserify/resolve/blob/75108bc1d196f0cf770c3c7777f349cc4c3373b2/lib/node-modules-paths.js#L12).

Does that sound right to you?

Turbo87 commented:
to be honest, I don't know. but when I discussed the implementation with @rwjblue a few weeks ago he seemed to have good reasons for the custom resolve options. I think it was related to Broccoli eventually using the system temp folder where naive node modules resolution will no longer work.

lennyburdette commented:
Interesting! @rwjblue didn't bring that up on the original ticket (#23) so I guess we'll need him to chime in here.

It does look like I missed another `projectRoot` reference related to caching: https://github.com/rwjblue/ember-cli-cjs-transform/blob/master/src/cjs-transform.js#L78

rwjblue commented on src/cjs-transform.js:
```diff
@@ -93,14 +96,18 @@ class CJSTransform extends Plugin {
+      fullPath = resolveSync(relativePath.slice(NODE_MODULES.length), {
+        basedir: this.projectRoot,
+      });
+    }
```
Can you add an `else` here that throws an error? I don't think we ever expect for `relativePath` to **not** include `node_modules/` do we?

lennyburdette replied:
good thinking! I switched it to throw early if we don't detect the "node_modules/" prefix.

lennyburdette commented:
@rwjblue I'm struggling to understand what `this.options` is in `calculateCacheKey` here:
```js
    for (let relativePath in this.options) {
      let fullPath = path.join(this.projectRoot, relativePath);
      let packageDir = pkgDir.sync(fullPath);
      let hash = hashForDep(packageDir);

      hashes.push(hash);
    }
```
Would this work?
```js
for (let relativePath in this.options) {
  let fullPath = resolveSync(relativePath.replace(/^node_modules\//, ''), { basedir: this.projectRoot });
  let packageDir = pkgDir.sync(fullPath);
  let hash = hashForDep(packageDir);

  hashes.push(hash);
}
```

lennyburdette commented:
Maybe the real question is: how do I write a test for this?

lennyburdette commented:
I added a test. The existing ones were actually very easy to read :grin:

rwjblue commented:
Discussed in ember-cli meeting today, this is looking really good.  Only small modification is to use `this.parent.root` instead of `this.project.root` for the base (in the hopes that this _somewhat_ mitigates issues with an addon having this as a direct dependency and doing `this.import`).

lennyburdette commented:
I'll try to capture the discussion before I was booted from the conference room:

1. This PR is an improvement on the current situation and is generally good-to-go.
2. For addons that use `this.import`, there is additional work to do to pass the correct `basedir` option to `resolve.sync`. It should be the addon's root directory in case they have a different version of a dependency (or we're in an "npm link" scenario). I believe this could be done in a later PR?

I'm not sure how to tackle number 2. Do we need to pass more information to the CJSTransform constructor here? (https://github.com/rwjblue/ember-cli-cjs-transform/blob/master/src/index.js#L18)

Commit:
fix: pass in `this.parent` instead of `this.project`

If the import transform is a dependency of an addon, *and* the addon has
a dependency mismatch with this host app, this should tell resolve to
find the addon's version of the dependency instead of the host app's.

lennyburdette commented:
Whoops, you're way faster than me—and I forgot the `this.parent` part!
